daemon/cmd/fqdn.go

```diff
@@ -180,13 +181,20 @@ func (d *Daemon) updateDNSDatapathRules(ctx context.Context) error {
 }
 
 // lookupEPByIP returns the endpoint that this IP belongs to
-func (d *Daemon) lookupEPByIP(endpointAddr netip.Addr) (endpoint *endpoint.Endpoint, err error) {
-	e := d.endpointManager.LookupIP(endpointAddr)
-	if e == nil {
-		return nil, fmt.Errorf("cannot find endpoint with IP %s", endpointAddr)
+func (d *Daemon) lookupEPByIP(endpointAddr netip.Addr) (endpoint *endpoint.Endpoint, isHost bool, err error) {
+	if e := d.endpointManager.LookupIP(endpointAddr); e != nil {
+		return e, e.IsHost(), nil
 	}
 
-	return e, nil
+	if node.IsNodeIP(endpointAddr) != "" {
+		if e := d.endpointManager.GetHostEndpoint(); e != nil {
```

Comment:
We do not currently support this behavior in L3/L4. You are creating a feature for L7 that does not exist for other layers. Please remove this from your PR and propose a way to make it work for L3/L4. [This PR](https://github.com/cilium/cilium/pull/27464) created a feature that made nodes selectable. You can use it as a template for how to make host selectable in L3/L4. Please bring this up in the #sig-policy channel on Cilium Slack or attend a weekly community meeting for further assistance.

Reply:
If I remove this, the whole feature will stop working, because then the DNS proxy will not be able to determine the Cilium endpoint ID corresponding to host's DNS request and will reject it, logging the message "Cannot extract endpoint ID from DNS request" (pkg/fqdn/dnsproxy/proxy.go:966). Host policy `toFQDNs` rules will then have nothing to work with.

> We do not currently support this behavior in L3/L4.

I do not understand what behavior you mean. Cilium already has the concept "this network flow comes from/to a host", doesn't it? Nodes (including local host) have endpoint IDs and security IDs, they can have network policies on egress and ingress. The only purpose of this part of the PR is to patch this knowledge through to DNS proxy. It does not care what those IP addresses are, it does not enable users to specify node addresses, and it does not expose host IP addresses anywhere they weren't visible or usable before. It is purely internal to Cilium agent. I only need to be able to associate a Cilium endpoint ID with DNS requests coming from the local node (host). If there is a way to do this without involving host IP addresses, which I admit is kludgy, I would be happy to do so. Perhaps DNS proxy could distinguish this by network interface the request came on (`cilium_host` vs anything else), although this would require extending miekg/dns to pass through network interface information in addition to address/port. Another alternative is to run DNS proxy on two ports and redirect host's DNS requests to a different proxy port than pods' DNS requests (this would require creating an extra pair of iptables -j TPROXY rules). The host-dedicated `ServeDNS` would not need to look up request's endpoint ID at all, it would just call `d.endpointManager.getHostEndpoint()` directly. This part of the PR does essentially the same thing.

Reply:
Resolved in this discussion on slack:
https://cilium.slack.com/archives/CDLN7836J/p1722360139578739